src/envoy/utils/utils.cc

```diff
@@ -120,30 +143,37 @@ bool GetDestinationUID(const envoy::api::v2::core::Metadata& metadata,
 
 bool GetPrincipal(const Network::Connection* connection, bool peer,
                   std::string* principal) {
-  if (connection) {
-    Ssl::ConnectionInfo* ssl =
-        const_cast<Ssl::ConnectionInfo*>(connection->ssl());
-    if (ssl != nullptr) {
-      const std::vector<std::string> sans =
-          (peer ? ssl->uriSanPeerCertificate() : ssl->uriSanLocalCertificate());
-      if (sans.empty()) {
-        // empty result is not allowed
-        return false;
-      }
-      const std::string result = sans[0];
-      if (result.length() >= kSPIFFEPrefix.length() &&
-          result.compare(0, kSPIFFEPrefix.length(), kSPIFFEPrefix) == 0) {
-        // Strip out the prefix "spiffe://" in the identity.
-        *principal = result.substr(kSPIFFEPrefix.size());
-      } else {
-        *principal = result;
-      }
-      return true;
+  std::string cert_san;
+  if (getCertSAN(connection, peer, &cert_san)) {
+    if (hasSPIFFEPrefix(cert_san)) {
+      // Strip out the prefix "spiffe://" in the identity.
+      *principal = cert_san.substr(kSPIFFEPrefix.size());
+    } else {
+      *principal = cert_san;
     }
+    return true;
   }
   return false;
 }
 
+bool GetTrustDomain(const Network::Connection* connection, bool peer,
+                    std::string* trust_domain) {
+  std::string cert_san;
+  if (!getCertSAN(connection, peer, &cert_san) || !hasSPIFFEPrefix(cert_san)) {
+    return false;
+  }
+
+  // Strip out the prefix "spiffe://" before getting trust domain.
+  std::size_t slash = cert_san.find('/', kSPIFFEPrefix.size());
```

Comment:
absl::StartsWith

Reply:
Is this comment for hasSPIFFEPrefix() or would you mind to clarify a little bit here? Thanks

Reply:
Aha, I totally misunderstand this one. I thought you are skipping "spiffe" instead of "spiffe://"
You are good to go. Sorry for the churn